sql: add randomized inverted index test

justinj opened the pull request:
This partially forwardports #23820 but the normalization rule is still
needed.

This test runs in about 2 seconds on my laptop, not sure if we'd rather
bring down the numbers significantly to try to cut that down.

Release note: None

Commit:
sql: add randomized inverted index test

This partially forwardports #23820 but the normalization rule is still
needed.

This test runs in about 2 seconds on my laptop, not sure if we'd rather
bring down the numbers significantly to try to cut that down.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24007)
<!-- Reviewable:end -->

jordanlewis commented on pkg/sql/tests/inverted_index_test.go:
```diff
@@ -0,0 +1,112 @@
+	db := sqlutils.MakeSQLRunner(tc.Conns[0])
+
+	db.Exec(t, "CREATE DATABASE IF NOT EXISTS test")
+	db.Exec(t, "CREATE TABLE test.jsons (i INT PRIMARY KEY, j JSONB, INVERTED INDEX (j))")
```
Can we insert half of the jsons before the inverted index was added, and half after?

justinj replied:
Great idea, done.

jordanlewis commented on pkg/sql/tests/inverted_index_test.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+func probeInvertedIndex(t *testing.T, db *sqlutils.SQLRunner, jsons []json.JSON) {
+	j := jsons[rand.Intn(len(jsons))]
```
You should use `rand.Perm` in the outer loop instead of this to avoid getting duplicates.

justinj replied:
Done, PTAL, made the test test a bunch more things.

jordanlewis reviewed (approved):
LGTM!

nit: I would use subtests here, since you're really testing 3-4 very different things. It would be nice to be able to see at a glance which part of this test was failing.

justinj commented:
Ok I:
* Added subtests
* Added a couple more rounds of testing: updating the primary key, doing deletions
* Do another check now that the number of returned results is correct - I spent some time breaking the implementation to make sure this caught regressions and it seemed not to without that change.

jordanlewis commented:
LGTM

justinj commented:
TFTR!
